src/Compilers/CSharp/Portable/Binder/InSubmissionClassBinder.cs

```diff
@@ -0,0 +1,83 @@
+    internal sealed class InSubmissionClassBinder : InContainerBinder
+    {
+        private readonly CompilationUnitSyntax _declarationSyntax;
+        private readonly bool _inUsings;
+        private QuickAttributeChecker? _lazyQuickAttributeChecker;
+
+        internal InSubmissionClassBinder(NamedTypeSymbol submissionClass, Binder next, CompilationUnitSyntax declarationSyntax, bool inUsings)
+            : base(submissionClass, next)
+        {
+            Debug.Assert(submissionClass.IsSubmissionClass);
+            _declarationSyntax = declarationSyntax;
+            _inUsings = inUsings;
+        }
+
+        internal override void GetCandidateExtensionMethods(
+            ArrayBuilder<MethodSymbol> methods,
+            string name,
+            int arity,
+            LookupOptions options,
+            Binder originalBinder)
+        {
+            for (var submission = this.Compilation; submission != null; submission = submission.PreviousSubmission)
+            {
+                submission.ScriptClass?.GetExtensionMethods(methods, name, arity, options);
+            }
+        }
+
+        internal override void LookupSymbolsInSingleBinder(
+            LookupResult result, string name, int arity, ConsList<TypeSymbol> basesBeingResolved, LookupOptions options, Binder originalBinder, bool diagnose, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            Debug.Assert(result.IsClear);
+
+            this.LookupMembersInSubmissions(result, (NamedTypeSymbol)Container, _declarationSyntax, _inUsings, name, arity, basesBeingResolved, options, originalBinder, diagnose, ref useSiteInfo);
+        }
+
+        protected override void AddLookupSymbolsInfoInSingleBinder(LookupSymbolsInfo result, LookupOptions options, Binder originalBinder)
+        {
+            this.AddMemberLookupSymbolsInfoInSubmissions(result, (NamedTypeSymbol)Container, _inUsings, options, originalBinder);
+        }
+
+        internal override ImmutableArray<AliasAndExternAliasDirective> ExternAliases => ((SourceNamespaceSymbol)Compilation.SourceModule.GlobalNamespace).GetExternAliases(_declarationSyntax);
+
+        internal override ImmutableArray<AliasAndUsingDirective> UsingAliases => ((SourceNamespaceSymbol)Compilation.SourceModule.GlobalNamespace).GetUsingAliases(_declarationSyntax, basesBeingResolved: null);
+
+        /// <summary>
+        /// Get <see cref="QuickAttributeChecker"/> that can be used to quickly
+        /// check for certain attribute applications in context of this binder.
+        /// </summary>
+        internal override QuickAttributeChecker QuickAttributeChecker
+        {
+            get
+            {
+                if (_lazyQuickAttributeChecker == null)
+                {
+                    QuickAttributeChecker result = this.Next!.QuickAttributeChecker;
+                    result = result.AddAliasesIfAny(_declarationSyntax.Usings);
+                    _lazyQuickAttributeChecker = result;
```

Comment:
Should this be an interlocked compareexchange? I don't think the data in the checker will be different if this was entered twice concurrently, but it could potentially return different instances of the checker itself.

Reply:
> Should this be an interlocked compareexchange? I don't think the data in the checker will be different if this was entered twice concurrently, but it could potentially return different instances of the checker itself.

This is insignificant for our purpose. The code is not an invention of this PR, it was moved from ImportsBinder.cs

---
In reply to: [588734372](https://github.com/dotnet/roslyn/pull/51676#discussion_r588734372) [](ancestors = 588734372)